libraries/HTTPClient/src/HTTPClient.cpp

```diff
@@ -919,21 +919,14 @@ int HTTPClient::writeToStream(Stream * stream)
  */
 String HTTPClient::getString(void)
 {
-    StreamString sstring;
-
-    if(_size > 0) {
-        // try to reserve needed memmory
-        if(!sstring.reserve((_size + 1))) {
-            log_d("not enough memory to reserve a string! need: %d", (_size + 1));
-            return "";
-        }
-    }
-    else {
-        return "";
+    // _size can be -1 when Server sends no Content-Length header
+    if(_size != 0) {
```

Comment:
if size can be -1 then this block may be unintentionally entered. I think you might want `if(_size > 0)`

Reply:
But that's exactly my use case (which from existing comment I assume is valid and not just some obscure problem on my side)  : size is -1 because the server hasn't put the relevant header in the response, BUT there is valid data to be read from the server so I do want to enter this block and read it.

Reply:
Maybe I should put `if(_size > 0 || _size == -1)` to be super explicit ??

Reply:
I'm not sure what server you are connecting to but nearly all do send the Content-Length header. I think your explicit check should be better.

Reply:
Another option is to remove the else clause and leave the if alone

Reply:
Google scripts (I use this from my ESP32 to pull data from Google Drive, etc. ...):
https://script.googleusercontent.com/macros/echo?user_content_key=g_uWqKQm9SRnYzPUr8zlBKhb4xJe2aMNKC_WtktO-5JD9ibF04HScAPl-4KWqei6ad0sK5CgJ1y5GGi6kjQ-_VujQEJrTSj1m5_BxDlH2jW0nuo2oDemN9CCS2h10ox_1xSncGQajx_ryfhECjZEnEXX6cJoH7fZw5b8BrSbrnico0hlOeVt3nbQiyiUYOk816bc8RKBUVWSsOSSCSTGmrxy_u9Yn2NR&lib=MBFYKdoFPkY3yJEvlYtZAhIyQU_TzVcy6

Re the "else" clause that's exactly what I did, I removed it:

```
    // _size can be -1 when Server sends no Content-Length header
    if(_size != 0) {
        StreamString sstring;
        writeToStream(&sstring);
        return sstring;
    }

    return "";
```

Reply:
I'd suggest revert the change and just remove the else block since the if block is used to reserve X bytes of memory before trying to read the page in. With your update it will do a number of re-allocations as the size of the stream increases which is not ideal when the size is known.

Reply:
Ok, fair enough, updated as suggested, with both  the more explicit `if(_size > 0 || _size == -1)` condition and the advance reservation of memory optimization.